rosbag2_cpp/src/rosbag2_cpp/info.cpp

```diff
@@ -67,9 +69,88 @@ struct service_req_resp_info
   std::unordered_map<client_id, sequence_set, client_id_hash> request;
   std::unordered_map<client_id, sequence_set, client_id_hash> response;
 };
+
+struct action_service_req_resp_info
+{
+  service_req_resp_info send_goal_service;
+  service_req_resp_info cancel_goal_service;
+  service_req_resp_info get_result_service;
+};
+
+inline void calculate_message_counts(
+  std::unordered_map<client_id, sequence_set, client_id_hash> & message_map,
+  size_t & count)
+{
+  count = 0;
+  for (auto & [client_id, message_list] : message_map) {
+    count += message_list.size();
+  }
+}
+
+using action_analysis =
+  std::unordered_map<std::string, std::shared_ptr<action_service_req_resp_info>>;
+
+inline void summary_action_service_info(
+  action_analysis & action_process_info,
+  std::unordered_map<std::string, std::shared_ptr<rosbag2_action_info_t>> & all_action_info)
+{
+  for (auto & [action_name, action_info] : action_process_info) {
+    size_t count = 0;
+    // Get the number of request from all clients for send_goal
+    calculate_message_counts(action_info->send_goal_service.request, count);
+    all_action_info[action_name]->send_goal_service_msg_count.first = count;
+
+    // Get the number of request from all clients for cancel_goal
+    count = 0;
+    calculate_message_counts(action_info->cancel_goal_service.request, count);
+    all_action_info[action_name]->cancel_goal_service_msg_count.first = count;
+
+    // Get the number of request from all clients for get_result
+    count = 0;
+    calculate_message_counts(action_info->get_result_service.request, count);
+    all_action_info[action_name]->get_result_service_msg_count.first = count;
+
+    // Get the number of response from all clients for send_goal
+    count = 0;
+    calculate_message_counts(action_info->send_goal_service.response, count);
+    all_action_info[action_name]->send_goal_service_msg_count.second = count;
+
+    // Get the number of response from all clients for cancel_goal
+    count = 0;
+    calculate_message_counts(action_info->cancel_goal_service.response, count);
+    all_action_info[action_name]->cancel_goal_service_msg_count.second = count;
+
+    // Get the number of response from all clients for get_result
+    count = 0;
+    calculate_message_counts(action_info->get_result_service.response, count);
+    all_action_info[action_name]->get_result_service_msg_count.second = count;
+  }
+}
+
+using service_analysis =
+  std::unordered_map<std::string, std::shared_ptr<service_req_resp_info>>;
+
+inline void summary_service_info(
```

Comment:
Please consider renaming this function for clarity
```suggestion
inline void update_service_info_with_num_req_resp(
```